README.md
# Current state of this repo

Previous versions of `fsharp-mode` (mis)used a second repository [emacs-fsharp-mode-bin](rneatherway/emacs-fsharp-mode-bin) to distribute `fsautocomplete.exe` and `fsharp-mode` together.

Future versions of `fsharp-mode` on [melpa](https://melpa.org/#/fsharp-mode) will be distributed directly from this repository and the installation of `fsautocomplete.exe` will
be done by the lsp-client [eglot](https://github.com/joaotavora/eglot) or [lsp-mode](https://github.com/emacs-lsp/lsp-mode).

On my build system all tests (using `eglot`) are passed but on travis and appveyor they fail because the tests now use .net core instead of mono: The .travis.yml configuration must be adapted.

If the tests pass through cleanly, I will change the [melpa recipe](https://melpa.org/#/fsharp-mode) to this repo.

# fsharp-mode

Provides support for the F# language in Emacs. Includes the following features:

- Support for F# Interactive
- Displays type signatures and tooltips
- Provides syntax highlighting and indentation.
- Intellisense support.

The following features are under development:

- Intelligent indentation

Requires Emacs 25.1+, Mono 3.10.X and F# 3.0 or higher. Without F# 3.0 (or higher)
the background compiler process will not function correctly.

## Build Status

### Travis [![Travis build status](https://travis-ci.org/fsharp/emacs-fsharp-mode.png)](https://travis-ci.org/fsharp/emacs-fsharp-mode)

See [.travis.yml](.travis.yml) for details.

## Installation

### Package

`fsharp-mode` is available on [MELPA](https://melpa.org) and can
be installed using the built-in package manager.

If you're not already using MELPA, add the following to your init.el:

```lisp
;;; Initialize MELPA
(require 'package)
(add-to-list 'package-archives '("melpa" . "http://melpa.org/packages/"))
(unless package-archive-contents (package-refresh-contents))
(package-initialize)

;;; Install fsharp-mode
(unless (package-installed-p 'fsharp-mode)
  (package-install 'fsharp-mode))

(require 'fsharp-mode)
```

If you are a user of [use-package](https://github.com/jwiegley/use-package) you can instead do

```lisp
(use-package fsharp-mode
  :defer t
  :ensure t)
```

All the [Configuration](#Configuration) options listed below can be added as part of the `:config` section on the `use-package` entry. 

### Manual installation

1. Clone this repo and run `make install`:
    ```
    git clone git://github.com/fsharp/emacs-fsharp-mode.git
    cd emacs-fsharp-mode/emacs
    make test-all # optional
    make install
    ```

2. Add the following to your init.el:
    ```lisp
    (add-to-list 'load-path "~/.emacs.d/fsharp-mode/")
    (autoload 'fsharp-mode "fsharp-mode"     "Major mode for editing F# code." t)
    (add-to-list 'auto-mode-alist '("\\.fs[iylx]?$" . fsharp-mode))
    ```

Note that if you do not use `make install`, which attempts to download
the dependencies from MELPA for you, then you must ensure that you have
installed them yourself. A list can be found in `fsharp-mode-pkg.el`.

If you run into any problems with installation, please check that you
have Emacs 24 on your PATH using `emacs --version`.
Note that OSX comes with Emacs 22 by default and installing a .app of
Emacs 24 will not add it to your PATH. One option is:

`alias emacs='/Applications/Emacs.app/Contents/MacOS/Emacs'`

## Usage

fsharp-mode should launch automatically whenever you open an F#
buffer. When the intellisense process is running, the following features will be available:

1. Type information for symbol at point will be displayed in the minibuffer.
2. Errors and warnings will be automatically highlighted, with mouseover
   text. Jump to the next and previous error using <kbd>M-n</kbd> and <kbd>M-p</kbd>.
3. To display a tooltip, move the cursor to a symbol and press
   <kbd>C-c C-t</kbd> (default).
4. To jump to the definition of a symbol at point, use <kbd>C-c C-d</kbd>.
5. Completion will be invoked automatically on dot, as in Visual Studio.
   It may be invoked manually using `fsharp-ac/complete-at-point`,
   bound by default to <kbd>C-c C-.</kbd>.
6. To stop the intellisense process for any reason, use <kbd>C-c C-q</kbd>.

### Projects

fsharp-mode offers intellisense for projects using the MSBuild/`.fsproj`
format. This allows project files to be shared with other developers using
Visual Studio and Xamarin Studio/Monodevelop. To create a new project file, it
is recommended that you take an existing project file and modify the list of
source files. One such project file can be found in the fsharp-mode repository
[here](https://github.com/fsharp/emacs-fsharp-mode/blob/master/test/Test1/Test1.fsproj). Alternatively
projects generated via [forge](https://github.com/ionide/Forge) are setup
with the needed files.

If, on loading a new `.fs` file, a `.fsproj` file is found in the
current or an enclosing directory, the intellisense process will be
launched if necessary, and the project loaded. When a project file is
loaded for the first time, this will be reported in the
minibuffer. Projects can also be manually loaded using <kbd>C-c
C-p</kbd>. This may be needed if a project file is in an unusual
location relative to a source file, or if the project file has been
modified (for example to add or remove a source file).

### Scripts

F# scripts (`.fsx` files) are standalone, and require no project file. If you wish open a script file and the intellisense process is not yet running, it will be launched automatically.

## Configuration

### Compiler and REPL paths

The F# compiler and interpreter should be set to good defaults for
your OS as long as the relevant executables can be found on your PATH
or in other standard locations. If you have a non-standard setup you
may need to configure these paths manually.

On Windows:

```lisp
(setq inferior-fsharp-program "\"c:\\Path\To\Fsi.exe\"")
(setq fsharp-compiler "\"c:\\Path\To\Fsc.exe\"")
```

On Unix-like systems, you must use the *--readline-* flag to ensure F#
Interactive will work correctly with Emacs. Typically `fsi` and `fsc` are
invoked through the shell scripts `fsharpi` and `fsharpc`.

```lisp
(setq inferior-fsharp-program "path/to/fsharpi --readline-")
(setq fsharp-compiler "path/to/fsharpc")
```

### Behavior

There are a few variables you can adjust to change how fsharp-mode behaves:

- `fsharp-ac-use-popup`: Show tooltips using a popup at the cursor
  position. If set to nil, display the tooltip in a split window.

- `fsharp-doc-idle-delay`: Set the time (in seconds) to wait before
  showing type information in the minibuffer.
  
- `fsharp-ac-intellisense-enabled`: This mode overrides some aspects of
  auto-complete configuration and runs the background process automatically.
  Set to nil to prevent this. Note that this will only prevent the background
  process from being launched in the *future*. If it is already running you
  will also need to quit it using <kbd>C-c C-q</kbd>.

You might also add `company-sort-prefer-same-case-prefix` to the list
of company transformers to to prefer exact case sort in completion
lists:

```lisp
(add-to-list 'company-transformers 'company-sort-prefer-same-case-prefix)
```

### Key Bindings

If you are new to Emacs, you might want to use the menu (call
menu-bar-mode if you don't see it). However, it's usually faster to learn
a few useful bindings:

 Key binding                         | Description
-------------------------------------|----------------------------------------
<kbd>C-c C-r</kbd>                   | Evaluate region
<kbd>C-c C-f</kbd>                   | Load current buffer into toplevel
<kbd>C-c C-e</kbd>                   | Evaluate current toplevel phrase
<kbd>C-M-x</kbd>                     | Evaluate current toplevel phrase
<kbd>C-M-h</kbd>                     | Mark current toplevel phrase
<kbd>C-c C-s</kbd>                   | Show interactive buffer
<kbd>C-c C-c</kbd>                   | Compile with fsc
<kbd>C-c x</kbd>                     | Run the executable
<kbd>C-c C-a</kbd>                   | Open alternate file (.fsi or .fs)
<kbd>C-c l</kbd>                     | Shift region to left
<kbd>C-c r</kbd>                     | Shift region to right
<kbd>C-c &lt;up&gt;</kbd>            | Move cursor to the beginning of the block
<kbd>C-c C-p</kbd>                   | Load a project for autocompletion and tooltips
<kbd>C-c C-d</kbd>, <kbd>M-.</kbd>   | Jump to definition of symbol at point
<kbd>C-c C-b</kbd>, <kbd>M-,</kbd>   | Return to where point was before jump.
<kbd>C-c C-t</kbd>                   | Request a tooltip for symbol at point
<kbd>C-c C-q</kbd>                   | Quit current background compiler process

To interrupt the interactive mode, use <kbd>C-c C-c</kbd>. This is useful if your
code does an infinite loop or a very long computation.

If you want to shift the region by 2 spaces, use: <kbd>M-2 C-c r</kbd>

In the interactive buffer, use <kbd>M-RET</kbd> to send the code without
explicitly adding the `;;` thing.

For key bindings that will be more familiar to users of Visual Studio, adding
the following to your `init.el` may be a good start:

```lisp
(add-hook 'fsharp-mode-hook
 (lambda ()
   (define-key fsharp-mode-map (kbd "M-RET") 'fsharp-eval-region)
   (define-key fsharp-mode-map (kbd "C-SPC") 'fsharp-ac/complete-at-point)))
```

## Editor

In order to change tab size it is possile to put this in emacs profile:

```lisp
(setq-default fsharp-indent-offset 2)
```

Because the F# language is sensitive to indentation, you might wan't to highlight indentation:

```lisp
(add-hook 'fsharp-mode-hook 'highlight-indentation-mode)
```

## Troubleshooting

`fsharp-mode` is still under development, so you may encounter some issues. Please report them so we can improve things! Either open an issue on [Github](https://github.com/fsharp/emacs-fsharp-mode/) with the label `Emacs`, or email the [mailing list](http://groups.google.com/group/fsharp-opensource).

### `fsharp-ac-debug`

If you set the variable `fsharp-ac-debug` to a non-`nil` value, e.g. `(setq fsharp-ac-debug 0)`, then some debug output will be seen in the buffer `*fsharp-debug*`. Setting `fsharp-ac-debug` to an 1 or 2 will cause a truncated or complete copy of communication between Emacs and the background intellisense process to be logged in `*fsharp-debug*`. This can make things rather slow, but would be useful for bug reports.

### Project file issues

If your project file does not seem to be being parsed correctly, so that you have missing references or other incorrect intellisense results, it is possible to obtain a detailed log of the project file loading process as follows:

* Open the F# file where the problems are visible.
* Set `fsharp-ac-debug` to 2, for example by `M-: (setq fsharp-ac-debug 2)`.
* Reload the project file using `C-c C-p`.
* Attach the detailed log output from your `*Messages*` buffer to a [new issue](https://github.com/fsharp/emacs-fsharp-mode/issues/new).

As an example, the log output will start similarly to:

```
<full path to>/MyProject.fsproj:
Loading default tasks for ToolsVersion: 4.0 from /usr/lib/mono/4.5/Microsoft.Common.tasks
<full path to>/MyProject.fsproj: Importing project
...
```

As an alternative, the command line tool responsible can be invoked directly to obtain the same log. Assuming you have installed from MELPA and are running on Linux or OS X, the following should work:

    mono ~/.emacs.d/elpa/fsharp-mode-<date>/bin/FSharp.Compiler.Service.ProjectCrackerTool.exe --text <path to>/MyProject.fsproj true

On Windows the `mono` prefix is not required, and if you have installed using a different method then the files will be in a different location.

### `Error: F# completion process produced malformed JSON.`


CHANGELOG.md
## 1.10 (2019.12-01)

Features:
  - #210: Remove old FsAutoComplete support (use LSP)
	- provide eglot (Emacs LSP client) integration and add eglot
      integration tests (using Emacs buttercup)
    - Use Cask instead to automate the package development cycle;
      development, dependencies, testing, building, packaging
    - Make project.el aware of F# projects
  - Use Emacs org-mode for README
Bugfixes:
    - #68: Indentation Cleanup / SMIE mode not being applied properly
      (Gastove)

## 1.9.14 (2019-06-09)

Features:
  - #207: Update to FsAutoComplete 0.38.1
  - #206: Set default build command to msbuild if found
Bugfixes:
  - #198: Use buffer-local version of company-quickhelp-mode

## 1.9.13 (2018)

Features:
  - #193: Update to FSAC 0.36
	 - Fixes #183: Load .Net Core projects that reference other projects
	 - Fixes #182: .fs files not parsed
Bugfixes:
  - #190: Fix attribute locking, improve imenu support
  - #189: Fix bug in font locking for active patterns
  - #187: Fix Infinite loop when file begins with a comment preceded
    by whitespace
  - #180: Use scoop instead of  Chocolatey package for Appveyor testing
  - #179: Use portable (Windows support) Makefile
  - #176: Add F# Tools 10.1 SDK directory to search dirs
  - #175: Paths with characters outside ASCII gives error FS2302 (Windows)
  - #171: Fix for phrase detection for if/then/else constructs

## 1.9.12 (2018-05-18)

Features:
  - #170: Flycheck verify (Improved fsautocomplete diagnostics)

Bugfixes:
  - #167: Fix error when visiting a new F# script file and
    fsautocomplete is not started
  - #168: Add Flycheck predicate function to prevent error when
    fsautocomplete is not running
  - #162: Stop matching [ as part of normal residue
  - #157: Don't change global value of `comment-indent-function'
  - #153: Add access control keywords to declaration regexes

## 1.9.11 (2017-10-21)

Features:
  - #151: Correctly find MSBuild from VS2017

Bugfixes:
  - #152: Handle failure to find build commands gracefully

## 1.9.10 (2017-09-18)

Bugfixes:
  - #146: Understand FSAC 0.34 error msgs

## 1.9.9 (2017-09-15)

Features
  - #143: Update to FsAutoComplete 0.34.0

Bugfixes:
  - #139: Disable flycheck and fsharp-doc-mode when fsharp-ac-intellisense-enabled is nil

## 1.9.8 (2017-06-17)

Features:
  - #134: Improved logging
  - #137: fsharp-shift-region-[left,right]: change bindings to 'C-c <' and 'C-c >'

Bugfixes:
  - #136: Use correct F# interactive prompt regex

## 1.9.7 (2017-06-06)

Bugfixes:
  - #131: Don't panic on malformed JSON (debug messages)
  - #133: Update faceup to capture font-locking <|

## 1.9.6 (2017-04-16)

Features:
  - #127: Update to FsAutoComplete 0.32.0 (.NET Core project support)

Bugfixes:
  - #125: Small fixes to try to prevent fsharp-mode to freeze all emacs 
  - #122: Make fsharp-doc-mode hook buffer-local

## 1.9.5 (2017-01-21)

Bugfixes:
  - #117: Fix `type` locking
  - #118: Don't change company-idle-delay
  - #120: Fix FSAC hanging issue

## 1.9.4 (2016-11-30)

Features
  - #116: Improve Active Pattern font locking, eval-when-compile the main font-lock-keywords form
  - #114: Clean up font-locking code

## 1.9.3 (2016-10-31)
Features
  - #111: Update to FsAutoComplete 0.32.0
  - #109: Define inferior-fsharp-mode as variant of comint mode

Bugfixes:
  - #110: Dont change default indent region function
  - #105: Don't send trailing newline to fsautocomplete
  - #104: Dont change `company-minimum-prefix-length'

## 1.9.2 (2016-09-30)
Features
  - #98: Enable imenu support

## 1.9.1 (2016-07-19)

Features:
  - Update to FsAutoComplete 0.29.0.

## 1.9.0 (2016-07-09)

Features:
  - #71: fontify the doc string (@nosami).
  - #77: Use new typesig command for fsharp-doc mode (@rneatherway).
  - #88: Use flycheck for error reporting (@juergenhoetzel).

Bugfixes:
  - #75: Do not change current buffer when starting FSI (@rneatherway).
  - #76: Record type highlighting (@rneatherway).
  - #79: Overlays should not grow when typing (@rneatherway).
  - #82: Inferior fsi: #silentcd to local directory in Tramp (@juergenhoetzel).
  - #83: Fix completion of type annotated symbols (@juergenhoetzel).
  - #85: Don't modify company-transformers (@nosami).
  - #86: Don't clobber company-backends (@nosami).

## 1.8.1 (2016-04-14)

Features:
  - #66: Tramp support (@juergenhoetzel).
  - #69: Prefer exact case sort in completion list (@nosami).

## 1.8.0 (2016-04-05)

Features:
  - Update to FsAutoComplete 0.28.0 to support #65.
  - #65: Faster completions (thanks to @nosami).
  - #56: Use FsAutoComplete "startswith" filter (thanks to @juergenhoetzel).

Bugfixes:
  - #67: Fix use of popup (thanks to @drvink)
  - #60: Unbreak company support on non-graphic displays (thanks to @drvink)
  - #58: Handle buffers not visiting a file (thanks to @juergenhoetzel).

## 1.7.4 (2016-02-05)

Features:
  - #49: Use company for completions (thanks to @nosami).

Bugfixes:
  - Update to FsAutoComplete 0.27.2, fixes project cracking for files
    with spaces in the path.

## 1.7.3 (2016-01-26)

Bugfixes:
  - Update to FsAutoComplete 0.27.1, fixes Windows VS2015-only support.

## 1.7.2 (2016-01-08)

Bugfixes:
  - #50: Inhibit electric-indent for fsharp-mode buffers (thanks to @joranvar).

## 1.7.1 (2015-11-24)

Features:
  - #45: Update FSAC to 0.27, enable project cracking logs.

## 1.7.0 (2015-11-24)

Features:
  - #34: Switch to SMIE-based indentation engine (thanks to m00nlight).
  - #31: Add highlighting of other usages of symbol at point.

## 1.6.3 (2015-10-24)

Bugfixes:
  - Update to FsAutoComplete 0.26.1, which fixes Windows support.

## 1.6.2 (2015-10-20)

Bugfixes:
  - Update to FsAutoComplete 0.26.0.
  - #30: Allow use of symbols containing '%'
  - #28: Fix FSI usage in buffers whose name differs from filename
  - #27: Fix test of fsharp-ac-debug

## 1.6.1 (2015-09-02)

Bugfixes:

fsharp-mode-font.el
;;; fsharp-mode-font.el --- Syntax highlighting for F#

;; Copyright (C) 1997 INRIA

;; Author: 1993-1997 Xavier Leroy, Jacques Garrigue and Ian T Zimmerman
;;         2010-2011 Laurent Le Brun <laurent@le-brun.eu>
;; Maintainer: Robin Neatherway <robin.neatherway@gmail.com>
;; Keywords: languages

;; This file is not part of GNU Emacs.

;; This file is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation; either version 3, or (at your option)
;; any later version.

;; This file is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with GNU Emacs; see the file COPYING.  If not, write to
;; the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
;; Boston, MA 02110-1301, USA.

;;; Commentary:

;;; Code:

;; (require 'fsharp-mode)
;; (require 'dash)

(defgroup fsharp-ui nil
  "F# UI group for the defcustom interface."
  :prefix "fsharp-ui-"
  :group 'fsharp
  :package-version '(fsharp-mode . "1.9.2"))

(defface fsharp-ui-generic-face
  '((t (:inherit default)))
  "Preprocessor face"
  :group 'fsharp-ui)

(defface fsharp-ui-operator-face
  '((t (:foreground "LightSkyBlue")))
  "Preprocessor face"
  :group 'fsharp-ui)

(defface fsharp-ui-warning-face
  '((t (:inherit font-lock-warning-face)))
  "Face for warnings."
  :group 'fsharp-ui)

(defface fsharp-ui-error-face
  '((t (:inherit font-lock-error-face :underline t)))
  "Face for errors"
  :group 'fsharp-ui)

(defmacro def-fsharp-compiled-var (sym init &optional docstring)
  "Defines a SYMBOL as a constant inside an eval-and-compile form
with initial value INITVALUE and optional DOCSTRING."
  `(eval-and-compile
     (defvar ,sym ,init ,docstring)))

(def-fsharp-compiled-var fsharp-shebang-regexp
  "\\(^#!.*?\\)\\([A-Za-z0-9_-]+\\)$"
  "Capture the #! and path of a shebag in one group and the
  executable in another.")

(def-fsharp-compiled-var fsharp-access-control-regexp
  "private\\s-+\\|internal\\s-+\\|public\\s-+"
  "Match `private', `internal', or `public', followed by a space,
  with no capture.")

(def-fsharp-compiled-var fsharp-access-control-regexp-noncapturing
  (format "\\(?:%s\\)" fsharp-access-control-regexp)
  "Same as `fsharp-access-control-regexp', but captures")

(def-fsharp-compiled-var fsharp-inline-rec-regexp
  "inline\\s-+\\|rec\\s-+"
  "Match `inline' or `rec', followed by a space.")

(def-fsharp-compiled-var fsharp-inline-rec-regexp-noncapturing
  (format "\\(?:%s\\)" fsharp-inline-rec-regexp)
  "Match `inline' or `rec', followed by a space, with no capture.")

(def-fsharp-compiled-var fsharp-valid-identifier-regexp
  "[A-Za-z0-9_']+"
  "Match a normal, valid F# identifier -- alphanumeric characters
  plus ' and underbar. Does not capture")

(def-fsharp-compiled-var fsharp-function-def-regexp
  (concat "\\<\\(?:let\\|and\\|with\\)\\s-+"
          fsharp-inline-rec-regexp-noncapturing "?"
          fsharp-access-control-regexp-noncapturing "*"
          (format "\\(%s\\)" fsharp-valid-identifier-regexp)
          "\\(?:\\s-+[A-Za-z_]\\|\\s-*(\\)" ;; matches function arguments or open-paren; unclear why 0-9 not in class
          ))

(def-fsharp-compiled-var fsharp-pattern-function-regexp
  (concat "\\<\\(?:let\\|and\\)\\s-+"
          fsharp-inline-rec-regexp-noncapturing "?"
          fsharp-access-control-regexp-noncapturing "*"
          (format "\\(%s\\)" fsharp-valid-identifier-regexp)
          "\\s-*=\\s-*function")
  "Matches an implicit matcher, eg let foo m = function | \"cat\" -> etc.")

;; Note that this regexp is used for iMenu. To font-lock active patterns, we
;; need to use an anchored match in fsharp-font-lock-keywords.
(def-fsharp-compiled-var fsharp-active-pattern-regexp
  (concat "\\<\\(?:let\\|and\\)\\s-+"
          fsharp-inline-rec-regexp-noncapturing "?"
          fsharp-access-control-regexp-noncapturing "*"
          "(\\(|[A-Za-z0-9_'|]+|\\))\\(?:\\s-+[A-Za-z_]\\|\\s-*(\\)"))

(def-fsharp-compiled-var fsharp-member-access-regexp
  "\\<\\(?:override\\|member\\|abstract\\)\\s-+"
  "Matches members declarations and modifiers on classes.")

(def-fsharp-compiled-var fsharp-member-function-regexp
  (concat fsharp-member-access-regexp
          fsharp-inline-rec-regexp-noncapturing "?"
          fsharp-access-control-regexp-noncapturing "*"
          "\\(?:" fsharp-valid-identifier-regexp "\\.\\)?"
          "\\(" fsharp-valid-identifier-regexp "\\)")
  "Captures the final identifier in a member function declaration.")

(def-fsharp-compiled-var fsharp-overload-operator-regexp
  (concat fsharp-member-access-regexp
          fsharp-inline-rec-regexp-noncapturing "?"
          fsharp-access-control-regexp-noncapturing "*"
          "\\(([!%&*+-./<=>?@^|~]+)\\)")
  "Match operators when overloaded by a type/class.")

(def-fsharp-compiled-var fsharp-constructor-regexp
  (concat "^\\s-*"
          fsharp-access-control-regexp-noncapturing "*"
          "\\<\\(new\\) *(.*)[^=]*=")
  "Matches the `new' keyword in a constructor")

(def-fsharp-compiled-var fsharp-type-def-regexp
  (concat "^\\s-*\\<\\(?:type\\|inherit\\)\\s-+"
          fsharp-access-control-regexp-noncapturing "*" ;; match access control 0 or more times
          "\\([A-Za-z0-9_'.]+\\)"))

(def-fsharp-compiled-var fsharp-var-or-arg-regexp
  "\\_<\\([A-Za-z_][A-Za-z0-9_']*\\)\\_>")

(def-fsharp-compiled-var fsharp-explicit-field-regexp
  (concat "^\\s-*\\(?:val\\|abstract\\)\\s-*\\(?:mutable\\s-+\\)?"
          fsharp-access-control-regexp-noncapturing "*" ;; match access control 0 or more times
          "\\([A-Za-z_][A-Za-z0-9_']*\\)\\s-*:\\s-*\\([A-Za-z_][A-Za-z0-9_'<> \t]*\\)"))

(def-fsharp-compiled-var fsharp-attributes-regexp
  "\\[<[A-Za-z0-9_]+>\\]"
  "Match attributes like [<EntryPoint>]")


;; F# makes extensive use of operators, many of which have some kind of
;; structural significance.
;;
;; In particular:
;; (| ... |)                 -- banana clips for Active Patterns (handled separately)
;; <@ ... @> and <@@ ... @@> -- quoted expressions
;; <| and |>                 -- left and right pipe (also <||, <|||, ||>, |||>)
;; << and >>                 -- function composition
;; |                         -- match / type expressions

(def-fsharp-compiled-var fsharp-operator-quote-regexp
  "\\(<@\\{1,2\\}\\)\\(?:.*\\)\\(@\\{1,2\\}>\\)"
  "Font lock <@/<@@ and @>/@@> operators.")

(def-fsharp-compiled-var fsharp-operator-pipe-regexp
  "<|\\{1,3\\}\\||\\{1,3\\}>"
  "Match the full range of pipe operators -- |>, ||>, |||>, etc.")

(def-fsharp-compiled-var fsharp-operator-case-regexp
  "\\s-+\\(|\\)[A-Za-z0-9_' ]"
  "Match literal | in contexts like match and type declarations.")

(defvar fsharp-imenu-generic-expression
  `((nil ,(concat "^\\s-*" fsharp-function-def-regexp) 1)
    (nil ,(concat "^\\s-*" fsharp-pattern-function-regexp) 1)
    (nil ,(concat "^\\s-*" fsharp-active-pattern-regexp) 1)
    (nil ,(concat "^\\s-*" fsharp-member-function-regexp) 1)
    (nil ,(concat "^\\s-*" fsharp-overload-operator-regexp) 1)
    (nil ,fsharp-constructor-regexp 1)
    (nil ,fsharp-type-def-regexp 1)
    )
  "Provide iMenu support through font-locking regexen.")

(defun fsharp-imenu-load-index ()
  "Hook up the provided regexen to enable imenu support."
  (setq imenu-generic-expression fsharp-imenu-generic-expression))

(add-hook 'fsharp-mode-hook #'fsharp-imenu-load-index)

(defun fsharp-var-pre-form ()
  (save-excursion
    (re-search-forward "\\(:\\s-*\\w[^)]*\\)?=" nil t)
    (match-beginning 0)))

(defun fsharp-fun-pre-form ()
  (save-excursion
    (search-forward "->")))

;; Preprocessor directives (3.3)
(def-fsharp-compiled-var fsharp-ui-preproessor-directives
  '("#if" "#else" "#endif" "#light"))

;; Compiler directives (12.4)
(def-fsharp-compiled-var fsharp-ui-compiler-directives
  '("#nowarn" "#load" "#r" "#reference" "#I"
    "#Include" "#q" "#quit" "#time" "#help"))

;; Lexical matters (18.4)
(def-fsharp-compiled-var fsharp-ui-lexical-matters
  '("#indent"))

;; Line Directives (3.9)
(def-fsharp-compiled-var fsharp-ui-line-directives
  '("#line"))

;; Identifier replacements (3.11)
(def-fsharp-compiled-var fsharp-ui-identifier-replacements
  '("__SOURCE_DIRECTORY__" "__SOURCE_FILE__" "__LINE__"))

;; F# keywords (3.4)
(def-fsharp-compiled-var fsharp-ui-fsharp-threefour-keywords
  '("abstract" "and" "as" "assert" "base" "begin"
    "class" "default" "delegate" "do" "do!" "done"
    "downcast" "downto" "elif" "else" "end"
    "exception" "extern" "false" "finally" "for" "fun"
    "function" "global" "if" "in" "inherit" "inline"
    "interface" "internal" "lazy" "let" "let!"
    "match" "member" "module" "mutable" "namespace"
    "new" "not" "null" "of" "open" "or" "override"
    "private" "public" "rec" "return" "return!"
    "select" "static" "struct" "then" "to" "true"
    "try" "type" "upcast" "use" "use!"  "val" "void"
    "when" "while" "with" "yield" "yield!"))

;; "Reserved because they are reserved in OCaml"
(def-fsharp-compiled-var fsharp-ui-ocaml-reserved-words
  '("asr" "land" "lor" "lsl" "lsr" "lxor" "mod" "sig"))

;; F# reserved words for future use
(def-fsharp-compiled-var fsharp-ui-reserved-words
  '("atomic" "break" "checked" "component" "const"
    "constraint" "constructor" "continue" "eager"
    "event" "external" "fixed" "functor" "include"
    "method" "mixin" "object" "parallel" "process"
    "protected" "pure" "sealed" "tailcall" "trait"
    "virtual" "volatile"))

;; RMD 2016-09-30 -- This was pulled out separately with the following comment
;; when I got here. Not clear to me why it's on it's own, or even precisely what
;; the comment means. But: `async' is a valid F# keyword and needs to go someplace,
;; so I've left it here. For now.
;;
;; Workflows not yet handled by fsautocomplete but async
;; always present
(def-fsharp-compiled-var fsharp-ui-async-words
  '("async")
  "Just the word async, in a list.")

(def-fsharp-compiled-var fsharp-ui-word-list-regexp
  (regexp-opt
   `(,@fsharp-ui-async-words
     ,@fsharp-ui-compiler-directives
     ,@fsharp-ui-fsharp-threefour-keywords
     ,@fsharp-ui-identifier-replacements
     ,@fsharp-ui-lexical-matters
     ,@fsharp-ui-ocaml-reserved-words
     ,@fsharp-ui-preproessor-directives
     ,@fsharp-ui-reserved-words
     ,@fsharp-ui-line-directives)
   'symbols))

(defconst fsharp-font-lock-keywords
  (eval-when-compile
    `((,fsharp-ui-word-list-regexp 0 font-lock-keyword-face)
      ;; shebang
      (,fsharp-shebang-regexp
       (1 font-lock-comment-face)
       (2 font-lock-keyword-face))
      ;; attributes
      (,fsharp-attributes-regexp . font-lock-preprocessor-face)
      ;; ;; type defines
      (,fsharp-type-def-regexp 1 font-lock-type-face)
      (,fsharp-function-def-regexp 1 font-lock-function-name-face)
      (,fsharp-pattern-function-regexp 1 font-lock-function-name-face)
      ;; Active Pattern
      ("(|" (0 'fsharp-ui-operator-face)
       ("\\([A-Za-z'_]+\\)\\(|)?\\)"
        nil nil
        (1 font-lock-function-name-face)
        (2 'fsharp-ui-operator-face)))
      (,fsharp-operator-pipe-regexp . 'fsharp-ui-operator-face)
      (,fsharp-member-function-regexp 1 font-lock-function-name-face)
      (,fsharp-overload-operator-regexp 1 font-lock-function-name-face)
      (,fsharp-constructor-regexp 1 font-lock-function-name-face)
      (,fsharp-operator-case-regexp 1 'fsharp-ui-operator-face)
      (,fsharp-operator-quote-regexp  (1 'fsharp-ui-operator-face)
                                      (2 'fsharp-ui-operator-face))
      ("[^:]:\\s-*\\(\\<[A-Za-z0-9_' ]*[^ ;\n,)}=<-]\\)\\(<[^>]*>\\)?"
       (1 font-lock-type-face)
       ;; 'prevent generic type arguments from being rendered in variable face
       (2 'fsharp-ui-generic-face nil t))
      (,(format "^\\s-*\\<\\(let\\|use\\|override\\|member\\|and\\|\\(?:%snew\\)\\)\\_>"
                (concat fsharp-access-control-regexp "*"))
       (0 font-lock-keyword-face) ; let binding and function arguments
       (,fsharp-var-or-arg-regexp
        (fsharp-var-pre-form) nil
        (1 font-lock-variable-name-face nil t)))
      ("\\<fun\\>"
       (0 font-lock-keyword-face) ; lambda function arguments
       (,fsharp-var-or-arg-regexp
        (fsharp-fun-pre-form) nil
        (1 font-lock-variable-name-face nil t)))
      (,fsharp-type-def-regexp
       (0 'font-lock-keyword-face) ; implicit constructor arguments
       (,fsharp-var-or-arg-regexp
        (fsharp-var-pre-form) nil
        (1 font-lock-variable-name-face nil t)))
      (,fsharp-explicit-field-regexp
       (1 font-lock-variable-name-face)
       (2 font-lock-type-face))

      ;; open namespace
      ("\\<open\s\\([A-Za-z0-9_.]+\\)" 1 font-lock-type-face)

      ;; module/namespace
      ("\\_<\\(?:module\\|namespace\\)\s\\([A-Za-z0-9_.]+\\)" 1 font-lock-type-face)
      )))

(defun fsharp-ui-setup-font-lock ()
  "Set up font locking for F# Mode."
  (setq font-lock-defaults
        '(fsharp-font-lock-keywords)))

(add-hook 'fsharp-mode-hook #'fsharp-ui-setup-font-lock)

(defun fsharp--syntax-propertize-function (start end)
  (goto-char start)
  (fsharp--syntax-string end)
  (funcall (syntax-propertize-rules
            ("\\(@\\)\"" (1 (prog1 "|" (fsharp--syntax-string end)))) ; verbatim string
            ("\\(\"\\)\"\"" (1 (prog1 "|" (fsharp--syntax-string end)))) ; triple-quoted string
            ("\\('\\)\\(?:[^\n\t\r\b\a\f\v\\\\]\\|\\\\[\"'ntrbafv\\\\]\\|\\\\u[0-9A-Fa-f]\\{4\\}\\|\\\\[0-9]\\{3\\}\\)\\('\\)"
             (1 "|") (2 "|")) ; character literal
            ("\\((\\)/" (1 "()"))
            ("\\(\(\\)\\*[!%&*+-\\./<=>@^|~?]*[\n\t\r\b\a\f\v ]*\)" (1 "()")) ; symbolic operator starting (* is not a comment
            ("\\(/\\)\\*" (1 ".")))
           start end))

(defun fsharp--syntax-string (end)
  (let* ((pst (syntax-ppss))
         (instr (nth 3 pst))
         (start (nth 8 pst)))
    (when (eq t instr) ; Then we are in a custom string
      (cond
       ((eq ?@ (char-after start)) ; Then we are in a verbatim string
        (while
            (when (re-search-forward "\"\"?" end 'move)
              (if (> (- (match-end 0) (match-beginning 0)) 1)
                  t ;; Skip this "" and keep looking further.
                (put-text-property (- (match-beginning 0) 1) (- (match-end 0) 1)
                                   'syntax-table (string-to-syntax "."))
                (put-text-property (match-beginning 0) (match-end 0)
                                   'syntax-table (string-to-syntax "|"))
                nil)))
        )

       (t ; Then we are in a triple-quoted string
        (when (re-search-forward "\"\"\"" end 'move)
          (put-text-property (- (match-beginning 0) 1) (match-beginning 0)
                             'syntax-table (string-to-syntax "."))
          (put-text-property (match-beginning 0) (match-end 0)
                             'syntax-table (string-to-syntax "|")))
        )))))

(provide 'fsharp-mode-font)

;;; fsharp-mode-font.el ends here
